File: C++/circular_buffer.cpp
#include<iostream>
#include<stdbool.h>

using namespace std;

#define MAX_LEN 3

template <typename T>
class circular_buffer{
    private:
        T buffer[MAX_LEN];
        int read_index, len, write_index;
        int capacity = MAX_LEN;

    public:
        circular_buffer()
        {
            read_index = 0;
            len = 0;
            write_index = 0;
        }
        void enqueue(T data)
        {
            if(isFull())
            {
                read_index = (read_index + 1) % capacity;
                cout << "Buffer is full" << endl;
            }
       
            buffer[write_index] = data;
            write_index = (write_index + 1) % capacity;
            if(len < capacity) len +=1;
        

            return;
        }

        T dequeue()
        {
            if(!isEmpty())
            {
                T data = buffer[read_index];
                read_index = (read_index + 1) % capacity;
                len -= 1;
                return data;
            }

            return T{};
        }

        void print_cb()
        {
            if(!isEmpty())
            {
                for(int i = 0; i < len; i++)
                {
                    int index = (read_index + i) % capacity;  // Correct circular indexing
                    cout << buffer[index] << " ";
                }
                cout << endl;
            }
        }


        bool isEmpty()
        {
            return len == 0;
        }

        bool isFull()
        {
            return len == capacity;
        }

};

int main()
{
    circular_buffer<int> cb;

    cb.enqueue(1);
    cb.enqueue(2);
    cb.enqueue(3);
    cb.dequeue();
    cb.print_cb();

    
}

File: Python/circular_buffer.py
class CircularBuffer:
    def __init__(self, max_len):
        self.max_len = max_len
        self.buffer = [None]*max_len
        self.write_Index = 0
        self.read_Index = 0
        self.count = 0
        
    def isFull(self):
        return self.max_len == self.count
    def isEmpty(self):
        return self.count == 0
    
    def enqueue(self, data):
        print(f"Read Index : {self.read_Index} Write Index : {self.write_Index} W_Data : {self.buffer[self.write_Index]} R_Data : {self.buffer[self.read_Index]}")
        if self.isFull():
            print("Buffer is full, overwriting the first element")
            self.read_Index = (self.read_Index + 1)% self.max_len
            
            
        self.buffer[self.write_Index] = data
        self.write_Index = (self.write_Index + 1) % self.max_len
        self.count = min(self.count + 1, self.max_len)
        print(f"Read Index : {self.read_Index} Write Index : {self.write_Index} W_Data : {self.buffer[self.write_Index]} R_Data : {self.buffer[self.read_Index]}")
        print()
        
    def dequeue(self):
        if self.isEmpty():
            print("buffer is empty")
            return
        
        data = self.buffer[self.read_Index]
        self.read_Index = (self.read_Index + 1)%self.read_Index
        self.count -=1
        return data
    
    def display(self):
        if self.isEmpty():
            print("Buffer is empty!")
            return

        print("Circular Buffer:", end=" ")
        index = self.read_Index
        for _ in range(self.count):
            print(self.buffer[index], end=" -> ")
            index = (index + 1) % self.max_len
        print("(End)")
        
cb = CircularBuffer(5)
cb.enqueue(1)
cb.enqueue(2)
cb.enqueue(3)
cb.enqueue(4)
cb.enqueue(5)
cb.enqueue(6)

cb.display()
    
            
           
    